modules/katello_content_view_promote.py

```diff
@@ -0,0 +1,148 @@
+    - "nailgun >= 0.29.0"
+    - "python >= 2.6"
+options:
+    server_url:
+        description:
+            - URL of Foreman server
+        required: true
+    username:
+        description:
+            - Username on Foreman server
+        required: true
+    password:
+        description:
+            - Password for user accessing Foreman server
+        required: true
+    verify_ssl:
+        description:
+            - Verify SSL of the Foreman server
+    name:
+        description:
+            - Name of the Katello content view
+        required: true
+    organization:
+        description:
+            - Organization that the content view is in
+        required: true
+    from_environment:
+        description:
+            - The lifecycle environment that is already associated with the content view version
+        required: true
+    to_environment:
+        description:
+            - The lifecycle environment to add to the content view version
+        required: true
+'''
+
+EXAMPLES = '''
+- name: "Promote the client content view from Library to Development"
+  katello_content_view_promote:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    verify_ssl: false
+    name: "Fedora CV"
+    organization: "Default Organization"
+    from_environment: "Library"
+    to_environment: "Development"
+'''
+
+RETURN = '''# '''
+
+try:
+    from ansible.module_utils.ansible_nailgun_cement import (
+        create_server,
+        handle_no_nailgun,
+        find_organization,
+        find_lifecycle_environment,
+        find_content_view,
+        find_content_view_version,
+        ping_server,
+    )
+    HAS_NAILGUN_PACKAGE = True
+except:
+    HAS_NAILGUN_PACKAGE = False
+
+
+def content_view_promote(module, name, organization, from_environment, to_environment):
+    changed = False
+
+    organization = find_organization(module, organization)
+    content_view = find_content_view(module, name=name, organization=organization)
+    from_environment = find_lifecycle_environment(module, name=from_environment, organization=organization)
+    to_environment = find_lifecycle_environment(module, name=to_environment, organization=organization)
+    content_view_version = find_content_view_version(module, content_view, from_environment)
+
+    if to_environment.id not in map(lambda environment: environment.id, content_view_version.environment):
+        content_view_version.promote(data={'environment_id': to_environment.id})
+        changed = True
+
+    return changed
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(type='bool', default=False),
+            name=dict(required=True),
+            organization=dict(required=True),
+            from_environment=dict(required=True),
```

Comment:
Do you need `from_environment`?
Looking at the http-post request, it seems katello only transfers the to_environment_id.

```
POST /katello/api/v2/content_view_versions/11/promote HTTP/1.1
[...]
{"id":11,"environment_id":10,"force":true}
```

Reply:
That's a fair question. This could be set to not be required and in that case promote from Library. That is not as explicit as specifying the `from_environment` which is required if you want to move whatever is in say Dev to Test.

Reply:
`from_environment` is used to find the content view version to promote to `to_environment`. Notice that it's used [on line 101](https://github.com/theforeman/foreman-ansible-modules/pull/38/files#diff-b584624a0fa1bd1a062f5161f39ac206R101).

Without `from_environment`, you would have to choose a version to promote. This might be fine if there's only one version, or if we assume the latest version. The current implementation matches the functionality in the [current ansible module](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/remote_management/foreman/katello.py#L400).

Reply:
Judging from the api itself [see below], my HTTP Headers must have com from an old version. ;)
the `environment_id` param is deprecated now.
But anyway when promoting a CV-version to an environment it does not matter where it came from. Right? And there seems to be no correspondence in the api.

```
    api :POST, "/content_view_versions/:id/promote", N_("Promote a content view version")
    param :id, :identifier, :desc => N_("Content view version identifier"), :required => true
    param :force, :bool, :desc => N_("force content view promotion and bypass lifecycle environment restriction")
    param :environment_id, :identifier, :deprecated => true, :desc => N_("LifeCycle Environment identifier")
    param :environment_ids, Array, :desc => N_("Identifiers for Lifecycle Environment")
    param :description, String, :desc => N_("The description for the content view version promotion")
    param :force_yum_metadata_regeneration, :bool, :desc => N_("Force metadata regeneration on the repositories " \
                                                           "in the content view version")
    def promote
      is_force = ::Foreman::Cast.to_bool(params[:force])
      task = async_task(::Actions::Katello::ContentView::Promote,
                        @version, @environments, is_force, params[:description],
                        :force_yum_metadata_regeneration => params[:force_yum_metadata_regeneration])
      respond_for_async :resource => task
    end
```

Reply:
In the API pasted, you can see the `param :id`, which is required. That is the content view version ID, which we find from the `from_environment` parameter.

Reply:
Ok, I see you can only have one Version in the from_environment. (This confuses me at all times. I would promote the environments to use specific versions, but that is katello's design decision.)
Imho, for the sake of idempotency, it would be better to specify the cv-version to be promoted, not the place it was supposed to come from.

Reply:
There is a check for idempotency. If an environment (i.e. env1) is promoted to another environment (i.e. env2), both environments will be associated with that version. If you run the playbook again, `changed` will be `False` because both env2 already exists on the version that env1 is associated with.